Add checks for test functions marked with @Test attribute in relevant Rules

hamtiko opened the pull request:
This pull request updates the `AlwaysUseLowerCamelCase` rule in the swift-format project to support the new [Swift Testing](https://github.com/apple/swift-testing) framework. 
Previously, the rule allowed underscores in test function names based on the old `XCTest` framework criteria. I’ve added a check to also allow underscores for functions marked with the `@Test` attribute.

P.S. I am planning to add check for new testing framework for `NeverForceUnwrap`,  `NeverUseForceTry`  and `NeverUseImplicitlyUnwrappedOptionals` rules as well.

allevato commented on Sources/SwiftFormat/Rules/AlwaysUseLowerCamelCase.swift:
```diff
@@ -100,7 +100,11 @@ public final class AlwaysUseLowerCamelCase: SyntaxLintRule {
     // We allow underscores in test names, because there's an existing convention of using
     // underscores to separate phrases in very detailed test names.
-    let allowUnderscores = testCaseFuncs.contains(node)
+    let allowUnderscores = testCaseFuncs.contains(node) || node.attributes.contains {
```
It would be great to factor this new check out into a convenience method on `WithAttributesSyntax`, which could then be used elsewhere (and on any declaration that can have attributes). What about something like this?

```
extension WithAttributesSyntax {
  func hasAttribute(_ name: String) { ... }
}
```

You could create a new `WithAttributesSyntax+Convenience.swift` file in Sources/SwiftFormat/Core for this (make sure to also add the new file to the CMakeLists.txt file in Sources/SwiftFormat).

If you're feeing really bold, you could also support the fully-qualified version: `@Testing.Test`.

hamtiko replied:
That's a good idea, will do it together with adding Swift Testing support for `NeverForceUnwrap`,  `NeverUseForceTry` and `NeverUseImplicitlyUnwrappedOptionals` rules to this PR. 
Thanks

allevato reviewed (commented):
> P.S. I am planning to add check for new testing framework for NeverForceUnwrap, NeverUseForceTry and NeverUseImplicitlyUnwrappedOptionals rules as well.

Would you be willing to do that in this PR? I don't think it would make it too big to review and it would be nice to have a single PR that adds both new swift-testing-related features.

hamtiko commented:
@allevato I have added the Swift Testing checks for `NeverForceUnwrap`, `NeverUseForceTry` and `NeverUseImplicitlyUnwrappedOptionals` rules to this PR and moved the attribute lookup logic to the separate extension as you suggested. Thanks

allevato commented on Sources/SwiftFormat/Core/WithAttributesSyntax+Convenience.swift:
```diff
@@ -0,0 +1,28 @@
+            let attributeName = attribute.as(AttributeSyntax.self)?.attributeName
+            return attributeName?.as(IdentifierTypeSyntax.self)?.name.text == name
+            // support @Module.Attribute syntax as well
+            || attributeName?.as(MemberTypeSyntax.self)?.name.text == name
```
Since you're not checking the module name here, this would match anything, right? Like `@Foo.Test`, for example.

It might be better to make the function signature `hasAttribute(_ name: String, inModule: String)` and have the caller always provide it, but it would only be checked if it was a `MemberTypeSyntax`. So `hasAttribute("Test", inModule: "Testing")` would still match either `@Test` or `@Testing.Test`, but not `@Foo.Test`.

hamtiko replied:
What you think if we will have 2 functions, one will be the current one  and another function like you suggested?

allevato replied:
I think it's better for simplicity to have one function that does both checks. We can't _definitively_ know if `@Test` refers to `@Testing.Test` without the file being type-checked, which we can't do in swift-format. So we're making an assumption that `@Test` by itself refers to the one in `Testing` and nothing else, because of how common we expect that pattern to be. If we're handling that as the common case, then there's no reason to not also handle `@Testing.Test` with the same query—each of the individual rules shouldn't have to make two calls to find the attribute it's looking for.

allevato commented on Tests/SwiftFormatTests/Rules/AlwaysUseLowerCamelCaseTests.swift:
```diff
@@ -210,4 +210,24 @@ final class AlwaysUseLowerCamelCaseTests: LintOrFormatRuleTestCase {
+      func 1️⃣function_Without_Test_Attribute() {}
+      @objc
+      func 2️⃣function_With_Non_Test_Attribute() {}
+      """,
```
Once you make the change above, you can add another example like `@Foo.Test` and verify that it does emit a diagnosis.

hamtiko replied:
Done.

allevato reviewed (commented):
Looks really good! Just one comment about the attribute checking.

allevato commented on Sources/SwiftFormat/Core/WithAttributesSyntax+Convenience.swift:
```diff
@@ -13,16 +13,23 @@
-            // support @Module.Attribute syntax as well
-            || attributeName?.as(MemberTypeSyntax.self)?.name.text == name
-        }
+  /// Indicates whether the node has attribute with the given `name`.
```
Can you mention in the doc comment that the module name is only considered if present if the attribute is written as `@Module.Name`, and that that's intentional? Otherwise it might look like a bug to someone reading it in the future.

hamtiko replied:
good point, done. 👍

allevato reviewed (approved):
Thank you for adding this!
